rust_src/src/decompress.rs

```diff
@@ -0,0 +1,134 @@
+use lisp::LispObject;
+use lisp::defsubr;
+use remacs_macros::lisp_fn;
+use remacs_sys::{buf_charpos_to_bytepos, del_range, insert_from_gap, make_gap, maybe_quit,
+                 move_gap_both};
+use threads::ThreadState;
+
+/// Return t if zlib decompression is available in this instance of Emacs.
+#[lisp_fn]
+pub fn zlib_available_p() -> bool {
+    true
+}
+
+enum CompressType {
+    Zlib,
+    Gzlib,
+    Raw,
+}
+
+fn detect_compress_type(magic_number: u8) -> CompressType {
+    match magic_number {
+        0x78 => CompressType::Zlib,
+        0x1F => CompressType::Gzlib,
+        // Assuming the data is raw , if can't find neither zlib nor gzib header.
+        _ => CompressType::Raw,
+    }
+}
+
+/// Decompress a gzip- or zlib-compressed region.
+/// Replace the text in the region by the decompressed data.
+/// On failure, return nil and leave the data in place.
+/// This function can be called only in unibyte buffers.
+#[lisp_fn]
+pub fn zlib_decompress_region(mut start: LispObject, mut end: LispObject) -> bool {
+    validate_region(&mut start, &mut end);
+
+    let mut current_buffer = ThreadState::current_buffer();
+
+    if current_buffer.multibyte_characters_enabled() {
+        error!("This function can be called only in unibyte buffers");
+    };
+
+    let istart = start.as_fixnum_or_error() as isize;
+    let iend = end.as_fixnum_or_error() as isize;
+
+    unsafe { move_gap_both(iend, iend) };
+
+    // Insert the decompressed data at the end of the compressed data.
+    let charpos = iend;
+    let bytepos = unsafe { buf_charpos_to_bytepos(current_buffer.as_mut(), iend as isize) };
+    let old_pt = current_buffer.pt();
+    current_buffer.set_pt_both(charpos, bytepos);
+
+    // Decompressed data
+    let byte_reader = unsafe {
+        slice::from_raw_parts(
+            current_buffer.byte_pos_addr(istart),
+            (iend - istart) as usize,
+        )
+    };
+
+    let magic_number = byte_reader[0];
+
+    let mut decoder: Box<Read> = match detect_compress_type(magic_number) {
+        CompressType::Zlib => Box::new(ZlibDecoder::new(byte_reader)),
+        CompressType::Gzlib => Box::new(GzDecoder::new(byte_reader)),
+        CompressType::Raw => Box::new(DeflateDecoder::new(byte_reader)),
+    };
+
+    let mut decompressed_bytes: isize = 0;
+
+    loop {
+        let avail_out: isize = 16 * 1024;
+
+        let gap_size = current_buffer.gap_size();
+
+        if gap_size < avail_out {
+            unsafe {
+                make_gap(avail_out - gap_size);
+            }
+        }
+
+        let gap_writer = unsafe {
+            slice::from_raw_parts_mut(current_buffer.gap_start_addr(), gap_size as usize)
+        };
+
+        match decoder.read(gap_writer) {
+            // Decompress finished.
+            Ok(0) => {
+                // Delete the compressed data.
+                unsafe { del_range(istart, iend) };
+                return true;
+            }
+
+            // Decompress once successfully.
```

Comment:
I don't entirely understand this comment -- could you rephrase it? I think at this point we've decoded the current data fine, but there's still some data left.